### PR TITLE
#799: Uploading schema with default value having integer breaks the schema

### PR DESCRIPTION
**Pull request by benedeki**
* Parsing the schema fields' metadata, that they are all of type String
* Reordering the database save operations, so that the parsing is done before
* Enhanced feedback info in case parsing fails

### CODE REVIEW

File: menas/src/main/scala/za/co/absa/enceladus/menas/utils/converters/SparkMenasSchemaConvertor.scala [hunk 1]
```diff
@@ -97,36 +100,60 @@ class SparkMenasSchemaConvertor @Autowired()(val objMapper: ObjectMapper) {
...
+
+    val childrenPath = SchemaUtils.appendPath(path, field.name)
+    val metadata: Map[String, String] = objMapper.readValue(field.metadata.json, classOf[Map[Any, Any]])
+        .collect{
+          // objMapper.readValue(_, classOf[Map[String, String]] doesn't actually enforce typing within the Map, this
+          // construction ensures it
+          case (key: String, value: String) => (key, value)
+          case (key, _) => throw SchemaParsingException(
+            schemaType = "",
+            message = s"Value for metadata key '$key' has to be a string",
```
[the remark this answers is not in the document]
Comment (benedeki) [reply menas/src/main/scala/za/co/absa/enceladus/menas/utils/converters/SparkMenasSchemaConvertor.scala]: Re 1: I tried in _Menas_ and it fails on decoding the uploaded Json. Will consider creating a test for it.

File: menas/src/main/scala/za/co/absa/enceladus/menas/utils/converters/SparkMenasSchemaConvertor.scala [hunk 2]
```diff
@@ -97,36 +100,60 @@ class SparkMenasSchemaConvertor @Autowired()(val objMapper: ObjectMapper) {
...
+  private def sparkToMenasField(field: StructField, path: String): SchemaField = {
+    val arr = field.dataType match {
+      case arrayType: ArrayType => Some(arrayType)
+      case _                    => None
+    }
+
+    val childrenPath = SchemaUtils.appendPath(path, field.name)
+    val metadata: Map[String, String] = objMapper.readValue(field.metadata.json, classOf[Map[Any, Any]])
+        .collect{
+          // objMapper.readValue(_, classOf[Map[String, String]] doesn't actually enforce typing within the Map, this
```
[the remark this answers is not in the document]
Comment (benedeki) [reply menas/src/main/scala/za/co/absa/enceladus/menas/utils/converters/SparkMenasSchemaConvertor.scala]: I guessed type erasure being he culprit of this as well.

File: menas/src/main/scala/za/co/absa/enceladus/menas/utils/converters/SparkMenasSchemaConvertor.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (benedeki) [reply menas/src/main/scala/za/co/absa/enceladus/menas/utils/converters/SparkMenasSchemaConvertor.scala]: Re 2: I can add it. But won't it potentially create a mess in the error message? The exception clearly states the field and location of the error... 🤔

[the remark this answers is not in the document]
Comment (benedeki) [reply menas/src/main/scala/za/co/absa/enceladus/menas/utils/converters/SparkMenasSchemaConvertor.scala]: Checked further on 1. If created within the code metadata require `String` to be the key when adding them. De-serialization can circumvent it maybe, but it's not a valid Json, for the key not to be a string. So IMO good enough this way.

[the remark this answers is not in the document]
Comment (benedeki) [reply menas/src/main/scala/za/co/absa/enceladus/menas/utils/converters/SparkMenasSchemaConvertor.scala]: Added the value to the error message. 😉

**Comment by GeorgiChochov** [comment]
Build seems to be failing (might be a problem after the branch is merged to develop)

**Comment by benedeki** [comment]
> 
> 
> Build seems to be failing (might be a problem after the branch is merged to develop)

nah. I changed the error message, but forgot to change the test checking the error accordingly. And committed thinking that nothing can go wring if just a string was changed. 🙄 (now verifying the fix is complete before committing and pushing, sorry)

**Comment by benedeki** [comment]
*For release notes:*
Metadata of fields in uploaded schema now allow only strings or `null` as values. Trying to upload a file with non-sting metadata doesn't break Menas anymore, instead properly reports the problem now.

**Pull request merged**